### ISSUE TITLE
Add support for adding ComicInfo.xml "inside" the cbz archive

### ISSUE THREAD

**Issue by Deivu** [HDoujinDownloader/HDoujinDownloader#110]
Some manga server like Komga wants the ComicInfo.xml inside the .cbz archive to load. Right now I manually add the ComicInfo.xml inside the archive, and it's very tedious specially on mangas that have a lot of chapters

**Comment by Tmp341** [comment]
I think, where available, there is a comicinfo.xml addition option on the app. 

Check: Settings>Save To>Metadata>File Formatting>Comicinfo.xml

**Comment by Deivu** [comment]
> I think, where available, there is a comicinfo.xml addition option on the app.
> 
> Check: Settings>Save To>Metadata>File Formatting>Comicinfo.xml

This generates the ComicInfo.xml beside the archives (inside the folder where the .cbz files are), which Komga refuses to load. It needs to be inside the .cbz archive instead, alongside with the images

**Comment by Tmp341** [comment]
Hmm. Normal txt infos are added inside the archive for me, so i assumed same occurs for comicinfo.xml too.

**Comment by Deivu** [comment]
Based on: https://github.com/gotson/komga/issues/866#issuecomment-1107377679
Seems like, the proper structure is to add the comicinfo.xml inside the `.cbz archive`. It would be very helpful for a toggle for that so users can choose to integrate the comicinfo.xml with basic info like tags description writer and title on it, pretty much the same comicinfo.xml that we generate, just an option to be it included on the cbz archive

**Comment by HDoujinDownloader** [comment]
@Tmp341 is correct that it should already be included in the archive by default, with some caveats:

For content that _doesn't_ have any chapters, the `ComicInfo.xml` file is always added to the archive. You'll end up with a CBZ archive that look like this:

```
/title_001.jpg
/title_002.jpg
...
/ComicInfo.xml
```

For content that _does_ have chapters, the result depends on the "Convert chapters individually" setting under _**Settings > Save to > Chapters**_.

If the setting is **enabled**, each chapter will be converted into a separate CBZ archive, with the `ComicInfo.xml` file sitting in the same directory as the chapter archives. It is not copied into any of the archives. (This is probably the problem you're encountering based on the issue you linked.)

If the setting is **disabled**, all the chapters will combined into a single CBZ archive, along with the `ComicInfo.xml` file. You'll end up with a CBZ archive that look like this:

```
/chapter 001/title_001.jpg
/chapter 002/title_002.jpg
...
/chapter 002/title_001.jpg
/chapter 002/title_002.jpg
...
/ComicInfo.xml
```

Some readers don't like when the archive contains subfolders like that, though. You can flatten it instead by enabling the "Download all chapters to single folder" setting under _**Settings > Save to > Chapters**_.

Hopefully that solves it! This issue does come up a lot, so maybe the default settings for this format could be improved.

**Comment by Deivu** [comment]
Thanks for the response, it did work, though does this mean if you convert chapters individually it wont contain a `comicinfo.xml`? If possible, is this something we could do something about like a toggle?

**Comment by HDoujinDownloader** [comment]
Yes, that is what it means-- But I'm definitely not opposed to adding something like this. We have some options:

- Simply copy the `ComicInfo.xml` into each chapter archive, or
- Generate a unique `ComicInfo.xml` for each chapter, with metadata pertaining to that chapter.

Maybe I could even add options for both, but I'm partial to the second one. Any opinion on this?

**Comment by Deivu** [comment]
I'm all good with the second one, where in we can default to the default `ComicInfo.xml` if no chapter info fetched. I feel like if we have both options, most users would just choose to generate the ComicInfo.xml, in this case, we will just have a fallback to use the default ComicInfo.xml when we can't generate a chapter unique ComicInfo.xml

**Comment by Tmp341** [comment]
>     * Generate a unique `ComicInfo.xml` for each chapter, with metadata pertaining to that chapter.
> 
So, this will work for both cases you mentioned above? (For content that has chapters and doesn't have chapters. Then it sounds more logical.

**Comment by HDoujinDownloader** [comment]
> So, this will work for both cases you mentioned above? (For content that has chapters and doesn't have chapters. Then it sounds more logical.

Yes, it'll work in both scenarios. 👍

**Comment by LuanaNight** [comment]
First off, I'm thrilled that you added ComicInfo.xml support, as I read my doujinshi mainly in Komga. But I think with some improvements it could be way better and more flexible.

Please add an option that allows the user to create their own mapping of the HDoujinDownloader fetched metadata fields, in relation to the ComicInfo.xml fields, so one can for example, have the artist name in the writer field instead of where it's currently put. It would also make it easier for the user to understand which fields will be written to where in the ComicInfo.xml file.

This would also remove the need for the Metadata.ComicInfo.PutTagsInGenre setting as the user could simply establish this mapping by themselves.

Due to the pre-made mapping, there is currently also a bug where under Metadata.General.IncludedFields, keeping the "Use default fields" checked vs unchecking it but keeping every sub check box checked, will remove many important tags from the ComicInfo.xml, like Series Group, which is very odd. This bug basically means that one currently can't uncheck any fields under Metadata.General.IncludedFields.

Cheers
And Happy Holidays

**Comment by HDoujinDownloader** [comment]
@LuanaNight Thanks a lot for your feedback on this-- I know there are a lot of opinions when it comes to how the metadata should be organized, and it's always helpful to get some new takes.

There was an oversight the last time I updated the ComicInfo.xml generator, and the artist information should be placed there by default with the next program update. The current priority for the writer field is _Author > Artist > Circle_, where the first non-empty value will be used. Let me know if this works for you, or if you'd like to see them prioritized differently (e.g. having the artist take top priority).

I'm wondering if that resolves your primary concern. I could allow the field mappings to be customized, but I feel like the majority of them would be typically left as-is.

Also, thanks for mentioning that bug. It's a result of the fact that if you un-check any fields, it saves the checked fields to a whitelist. Not every ComicInfo.xml field is listed there, so they end up disabled. I'll have this fixed with the next program update.

Happy holidays, by the way! 🎄
